kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -594,21 +639,26 @@ private StreamThreadProcessorContext.PreviousRecordContextAndNode preFinalize(
           event.partition(),
           taskId.toString()));
     }
-    final Optional<RuntimeException> processingException = event.processingException();
-    if (processingException.isPresent()) {
+
+    final var toClose = streamThreadContext.prepareToFinalizeEvent(event);
+
+    final Optional<Throwable> processingException = event.processingException();
+    if (processingException.isEmpty()) {
+      event.transitionToFinalizing();
+    } else {
       pendingEvents.remove(event);
-      this.fatalException = processingException.get();
-      throw this.fatalException;
+      event.transitionToDone();
```

Comment:
this part is different: not just that we don't set the `fatalException` here anymore, but we also transition events that were FAILED to DONE. Two reasons:
- I think it makes sense to retain DONE as the terminal state that's associated with an event being removed from the `pendingEvents` map
- We can thus use the event's state to make sure that we have checked on an error for a given event. If it's stuck in FAILED that means we haven't ever picked up the error and handled it on the StreamThread end. Once we do this, the event is transitioned from FAILED to DONE. So FAILED is  like an alternative to finalization rather than a terminal state of its own